homeassistant/components/lightwave.py

```diff
@@ -0,0 +1,153 @@
+"""
+homeassistant.components.lightwave.
+
+Implements communication with LightwaveRF.
+
+My understanding of the LightWave Hub is that devices cannot be discovered
+so must be registered manually. This is done in the configuration file.
+
+lightwave:
+    host: ip_address
+
+Where ip_address is the ip address of your LightwaveRF hub
+
+For more details on the api see: https://api.lightwaverf.com/
+"""
+import queue
+import threading
+import socket
+import time
+import logging
+import voluptuous as vol
+
+from homeassistant.const import CONF_HOST
+from homeassistant.helpers import config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+LIGHTWAVE_LINK = 'lightwave_link'
+DOMAIN = 'lightwave'
+LWRF_REGISTRATION = '100,!F*p'
+LWRF_DEREGISTRATION = '100,!F*xP'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Try to start embedded Lightwave broker."""
+    host = config[DOMAIN].get(CONF_HOST)
+    hass.data[LIGHTWAVE_LINK] = LWLink(host)
+    return True
+
+
+class LWLink():
+    """LWLink provides a communication link with the LightwaveRF hub."""
+
+    SOCKET_TIMEOUT = 2.0
+    RX_PORT = 9761
+    TX_PORT = 9760
+
+    the_queue = queue.Queue()
+    thread = None
+    link_ip = ''
+
+    def __init__(self, link_ip=None):
+        """Initialise the component."""
+        if link_ip is not None:
+            LWLink.link_ip = link_ip
+
+    def _send_message(self, msg):
+        """Add message to queue and start processing the queue."""
+        LWLink.the_queue.put_nowait(msg)
+        if LWLink.thread is None or not self.thread.isAlive():
+            LWLink.thread = threading.Thread(target=self._send_queue)
+            LWLink.thread.start()
+
+    def turn_on_light(self, device_id, name):
+        """Create the message to turn light on."""
+        msg = '321,!%sFdP32|Turn On|%s' % (device_id, name)
+        self._send_message(msg)
+
+    def turn_on_switch(self, device_id, name):
+        """Create the message to turn switch on."""
+        msg = '321,!%sF1|Turn On|%s' % (device_id, name)
+        self._send_message(msg)
+
+    def turn_on_with_brightness(self, device_id, name, brightness):
+        """Scale brightness from 0..255 to 1..32."""
+        brightness_value = round((brightness * 31) / 255) + 1
+        # F1 = Light on and F0 = light off. FdP[0..32] is brightness. 32 is
+        # full. We want that when turning the light on.
+        msg = '321,!%sFdP%d|Lights %d|%s' % (
+            device_id, brightness_value, brightness_value, name)
+        self._send_message(msg)
```

Comment:
Device-specific code must be move to a separate Python module and then imported.

Reply:
Now sure what to here. I am guessing apart from the setup code the entire contents of LWLink should move to a separate Python module. As I am new to open source how would I do this and where would it be published? Sorry for having to ask.

Reply:
Done.